gvm.go

```diff
@@ -1,163 +1,134 @@
-common shells.
-
-  bash:
-    eval "$(gvm 1.10)"
-
-  batch (windows cmd.exe):
-    FOR /f "tokens=*" %i IN ('"gvm.exe" 1.10') DO %i
-
-  powershell:
-    gvm --format=powershell 1.10 | Invoke-Expression
-`
+type Manager struct {
+	// GVM Home directory. Defaults to $HOME/.gvm
+	Home string
 
-// Output formats.
-const (
-	BashFormat       = "bash"
-	BatchFormat      = "batch"
-	PowershellFormat = "powershell"
-)
-
-var (
-	version = "SNAPSHOT"
+	// GOOS settings. Defaults to current OS.
+	GOOS string
 
-	log = logrus.WithField("package", "main")
-)
+	// GOARCH setting. Defaults to the current architecture.
+	GOARCH string
 
-type GVM struct {
-	Version      string
-	UseProjectGo bool
-	Format       string
+	// Golang binary store URL. Used to download listing and go binaries.
+	// Defaults to https://storage.googleapis.com/golang
+	GoStorageHome string
 
-	out io.Writer // Stdout writer (used for capturing output in tests).
+	cacheDir    string
+	versionsDir string
 }
 
-func (g *GVM) Run(_ *kingpin.ParseContext) error {
-	version := g.Version
-	if g.UseProjectGo {
-		ver, err := getProjectGoVersion()
+func (m *Manager) Init() error {
+	if m.Home == "" {
+		home, err := homeDir()
 		if err != nil {
 			return err
 		}
-		version = ver
-	}
 
-	if version == "" {
-		return fmt.Errorf("no version specified")
+		m.Home = filepath.Join(home, ".gvm")
 	}
-	log.Debugf("Using Go version %v", version)
 
-	goroot, err := golang.SetupGolang(version)
-	if err != nil {
-		return err
+	if m.GoStorageHome == "" {
+		m.GoStorageHome = "https://storage.googleapis.com/golang"
 	}
 
-	switch g.Format {
-	case BashFormat:
-		fmt.Fprintf(g.out, `export GOROOT="%v"`+"\n", goroot)
-		fmt.Fprintf(g.out, `export PATH="$GOROOT/bin:$PATH"`+"\n")
-		if strings.HasPrefix(version, "1.5") {
-			fmt.Fprintln(g.out, `export GO15VENDOREXPERIMENT=1`)
-		}
-	case BatchFormat:
-		fmt.Fprintf(g.out, `set GOROOT=%v`+"\n", goroot)
-		fmt.Fprintf(g.out, `set PATH=%s\bin;%s`+"\n", goroot, os.Getenv("PATH"))
-		if strings.HasPrefix(version, "1.5") {
-			fmt.Fprintln(g.out, `set GO15VENDOREXPERIMENT=1`)
-		}
-	case PowershellFormat:
-		fmt.Fprintf(g.out, `$env:GOROOT = "%v"`+"\n", goroot)
-		fmt.Fprintf(g.out, `$env:PATH = "$env:GOROOT\bin;$env:PATH"`+"\n")
-		if strings.HasPrefix(version, "1.5") {
-			fmt.Fprintln(g.out, `$env:GO15VENDOREXPERIMENT=1`)
+	if m.GOOS == "" {
+		m.GOOS = runtime.GOOS
+	}
+	if m.GOARCH == "" {
+		switch runtime.GOARCH {
+		default:
+			m.GOARCH = runtime.GOARCH
+		case "arm":
+			// The only binary releases are for ARM v6.
+			m.GOARCH = "armv6l"
 		}
-	default:
-		return errors.Errorf("invalid format option '%v'", g.Format)
 	}
 
+	m.cacheDir = filepath.Join(m.Home, "cache")
+	m.versionsDir = filepath.Join(m.Home, "versions")
+	return m.ensureDirStruct()
+}
+
+func (m *Manager) ensureDirStruct() error {
+	for _, dir := range []string{m.cacheDir, m.versionsDir} {
+		if err := os.MkdirAll(dir, os.ModeDir|os.ModePerm); err != nil {
```

Comment:
`os.ModeDir|os.ModePerm` - That's like 0777, right? That's too liberal for security purposes. I suggest `0755`.